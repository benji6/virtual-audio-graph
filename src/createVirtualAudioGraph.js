/* global AudioContext */
import {equals, forEach} from './utils'
import connectAudioNodes from './connectAudioNodes'
import createVirtualAudioNode from './createVirtualAudioNode'

const disconnectParents = (virtualNode, virtualNodes) => forEach(
  key => virtualNodes[key].disconnect(virtualNode),
  Object.keys(virtualNodes)
)

export default ({
  audioContext = new AudioContext(),
  output = audioContext.destination,
} = {}) => {
  return {
    audioContext,
    get currentTime () { return audioContext.currentTime },
    getAudioNodeById (id) { return this.virtualNodes[id].audioNode },
    update (newGraph) {
      forEach(id => {
        if (newGraph.hasOwnProperty(id)) return
        const virtualAudioNode = this.virtualNodes[id]
        virtualAudioNode.disconnectAndDestroy()
        disconnectParents(virtualAudioNode, this.virtualNodes)
        delete this.virtualNodes[id]
      }, Object.keys(this.virtualNodes))

      forEach(key => {
        if (key === 'output') throw new Error('"output" is not a valid id')
        const newNodeParams = newGraph[key]
        const [paramsNodeName, paramsOutput, paramsParams] = newNodeParams
        const virtualAudioNode = this.virtualNodes[key]
        if (virtualAudioNode == null) {
          this.virtualNodes[key] = createVirtualAudioNode(audioContext, newNodeParams)
          return
        }
        if (
          (paramsParams && paramsParams.startTime) !==
            (virtualAudioNode.params && virtualAudioNode.params.startTime) ||
          (paramsParams && paramsParams.stopTime) !==
            (virtualAudioNode.params && virtualAudioNode.params.stopTime) ||
          paramsNodeName !== virtualAudioNode.node
        ) {
          virtualAudioNode.disconnectAndDestroy()
          disconnectParents(virtualAudioNode, this.virtualNodes)
          this.virtualNodes[key] = createVirtualAudioNode(audioContext, newNodeParams)
          return
        }
        if (!equals(paramsOutput, virtualAudioNode.output)) {
          virtualAudioNode.disconnect()
          disconnectParents(virtualAudioNode, this.virtualNodes)
          virtualAudioNode.output = paramsOutput
        }

        virtualAudioNode.update(paramsParams)
      }, Object.keys(newGraph))

      connectAudioNodes(
        this.virtualNodes,
        virtualNode => virtualNode.connect(output)
      )

      return this
    },
    virtualNodes: {},
  }
}
