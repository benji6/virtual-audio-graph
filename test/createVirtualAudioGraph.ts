import createVirtualAudioGraph, * as V from "../src";

const audioContext: any = new AudioContext();

describe("createVirtualAudioGraph", () => {
  test("optionally takes audioContext property", () => {
    expect(createVirtualAudioGraph({ audioContext }).audioContext).toBe(
      audioContext,
    );
    expect(createVirtualAudioGraph().audioContext).not.toBe(audioContext);
    expect(createVirtualAudioGraph().audioContext).toBeInstanceOf(AudioContext);
  });

  test("optionally takes output parameter", () => {
    const gain: any = audioContext.createGain();

    createVirtualAudioGraph({ audioContext, output: gain }).update({
      0: V.gain(V.OUTPUT, { gain: 0.2 }),
    });

    expect(gain.toJSON()).toMatchSnapshot();

    createVirtualAudioGraph({ audioContext }).update({
      0: V.gain(V.OUTPUT, { gain: 0.2 }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });
});
