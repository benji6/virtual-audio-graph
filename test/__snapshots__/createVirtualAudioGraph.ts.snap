// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createVirtualAudioGraph optionally takes output parameter 1`] = `
{
  "gain": {
    "inputs": [],
    "value": 1,
  },
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.2,
      },
      "inputs": [],
      "name": "GainNode",
    },
  ],
  "name": "GainNode",
}
`;

exports[`createVirtualAudioGraph optionally takes output parameter 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.2,
      },
      "inputs": [],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;
